File: src/bumperbot_controllers/src/twist_relay.cpp
#include <bumperbot_controllers/twist_relay.hpp>

using std::placeholders::_1;
using namespace geometry_msgs::msg;

TwistRelay::TwistRelay(const std::string &name) : Node(name)
{
    declare_parameter("twist_in_topic", "bumperbot_controller/cmd_vel_unstamped");
    declare_parameter("twist_stamped_out_topic", "bumperbot_controller/cmd_vel");
    declare_parameter("twist_stamped_joy_in_topic", "input_joy/cmd_vel_stamped");
    declare_parameter("twist_joy_out_topic", "input_joy/cmd_vel");

    std::string twist_in_topic;
    std::string twist_stamped_out_topic;
    std::string twist_stamped_joy_in_topic;
    std::string twist_joy_out_topic;

    get_parameter("twist_in_topic", twist_in_topic);
    get_parameter("twist_stamped_out_topic", twist_stamped_out_topic);
    get_parameter("twist_stamped_joy_in_topic", twist_stamped_joy_in_topic);
    get_parameter("twist_joy_out_topic", twist_joy_out_topic);

    twist_sub_ = create_subscription<Twist>(twist_in_topic, 10, std::bind(&TwistRelay::twistCallback, this, _1));
    twist_pub_ = create_publisher<TwistStamped>(twist_stamped_out_topic, 10);
    joy_sub_ = create_subscription<TwistStamped>(twist_stamped_joy_in_topic, 10, std::bind(&TwistRelay::joyCallback, this, _1));
    joy_pub_ = create_publisher<Twist>(twist_joy_out_topic, 10);
}

void TwistRelay::twistCallback(const Twist & msg){
    TwistStamped twist_stamped;
    twist_stamped.header.stamp = get_clock()->now();
    twist_stamped.twist = msg;

    twist_pub_->publish(twist_stamped);
}

void TwistRelay::joyCallback(const TwistStamped & msg){
    Twist twist;
    twist.angular = msg.twist.angular;
    twist.linear = msg.twist.linear;

    joy_pub_->publish(twist);
}


int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TwistRelay>("twist_relay");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/bumperbot_controllers/bumperbot_controllers/twist_relay.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped, Twist

class TwistRelay(Node):
    def __init__(self):
        super().__init__("twist_relay")
        self.declare_parameter("twist_in_topic", "bumperbot_controller/cmd_vel_unstamped")
        self.declare_parameter("twist_stamped_out_topic", "bumperbot_controller/cmd_vel")
        self.declare_parameter("twist_stamped_joy_in_topic", "input_joy/cmd_vel_stamped")
        self.declare_parameter("twist_joy_out_topic", "input_joy/cmd_vel")
        
        self.twist_sub = self.create_subscription(
            Twist,
            self.get_parameter("twist_in_topic").get_parameter_value().string_value,
            self.twist_callback,
            10,
        )

        self.twist_pub = self.create_publisher(
            TwistStamped, 
            self.get_parameter("twist_stamped_out_topic").get_parameter_value().string_value, 
            10
        )

        self.joy_sub = self.create_subscription(
            TwistStamped,
            self.get_parameter("twist_stamped_joy_in_topic").get_parameter_value().string_value,
            self.joy_callback,
            10,
        )

        self.joy_pub = self.create_publisher(
            Twist,
            self.get_parameter("twist_joy_out_topic").get_parameter_value().string_value,
            10,
        )

        
    def twist_callback(self, msg):
        # Create a new TwistStamped message
        twist_stamped_msg = TwistStamped()
        twist_stamped_msg.header.stamp = self.get_clock().now().to_msg()
        twist_stamped_msg.twist = msg

        # Publish the TwistStamped message
        self.twist_pub.publish(twist_stamped_msg)

    def joy_callback(self, msg):
        # Create a new Twist message
        twist_msg = Twist()
        twist_msg.linear = msg.twist.linear
        twist_msg.angular = msg.twist.angular

        # Publish the Twist message
        self.joy_pub.publish(twist_msg)

def main(args=None):
    rclpy.init(args=args)
    twist_relay = TwistRelay()
    rclpy.spin(twist_relay)
    twist_relay.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()